### ISSUE TITLE
VM Agents do not come online up after latest Azure VM Agent plugin

### ISSUE THREAD

**Issue by ryanberger-az** [jenkinsci/azure-vm-agents-plugin#328]
### Version report

Jenkins and plugins versions report:

Jenkins: 2.321
Azure VM Agent: 799.va4c741108611

- What Operating System are you using (both controller, and any agents involved in the problem)?

Ubuntu for the controller, Windows Server 2019 for the agents.

Issue: Nodes are not fully deploying. Intermittently we see the resources being deployed into the resource group, but never checking into Jenkins as an agent. Build operations starts, but seems ultra delayed and log gets flooded with this message. Node never checks in to Jenkins. We are seeing this on multiple environments. **We are using Managed Identity to authenticate to Azure from our Jenkins controller.**

![image](https://user-images.githubusercontent.com/60238330/142477078-ba4a9a61-b4a0-46fe-94c4-1dc67470d262.png)

I have worked with @timja on a similar issue and am wondering if this has come back again. I noticed that the drop downs (resource group name, virtual machine size, region, etc) that are queries with API calls - never receive an answer. They just sit indefinitely spinning.

We saw this same behavior last time and I believe changes were made to resolve this. I noticed that if I click inside of the drop down, it shows a value (the correct one) but it never populates. This behavior is the same for all 3 drop downs (resource group name, virtual machine size, region, etc).

![image](https://user-images.githubusercontent.com/60238330/142477693-d3717439-e3a0-41c1-af73-ab959d4a9405.png)

![image](https://user-images.githubusercontent.com/60238330/142477756-5856d111-f889-4800-a151-4ecc89e87cf9.png)

In short, ever since updating to the newest version of the Azure VM Agent plugin - I can no longer successfully spin up VM Agents. Since I know these plugins are tied together, here is a snapshot of all of the versions we're running.

![image](https://user-images.githubusercontent.com/60238330/142477919-937937b8-9faa-4573-b7b9-70369a382160.png)

**Comment by ryanberger-az** [comment]
It looks like the API calls are taking a ridiculous amount of time to run. I left a page idle for 5-10 minutes and then peaked back in and it did fill in one of the drop downs. The others are still spinning.

**Comment by timja** [comment]
Is this perhaps to do with updating Jenkins rather than the plugin?

**Comment by ryanberger-az** [comment]
I rolled back to 2.316 and the issue persists. Same behavior with the drop downs not populating and nodes not properly deploying (or even starting to deploy from what it seems).

**Comment by timja** [comment]
K unlikely to be an issue with the latest release, maybe something in the azure-sdk plugin

**Comment by ryanberger-az** [comment]
Do I need to make an issue somewhere else then? If that is the case?

**Comment by timja** [comment]
Not without more information

**Comment by ryanberger-az** [comment]
Hi @timja - I am more than happy to get you whatever you would need or if you point me in the right direction. I will take the issue there. The only reason I reported it here is because I have worked with you on a very similar issue before, and thought it may be related.

**Comment by timja** [comment]
Could you try hit the MSI identity from the machine that Jenkins is running on and make sure it’s responding ok.

You could also try with an SP as well possibly

**Comment by ryanberger-az** [comment]
I tried to toggle on/off the managed identity from the VM. I then re-made the credential in Jenkins, reassociated the cloud configuration with that new managed identity. That filled in the resource group name, properly. The other fields are perpetually spinning. I tried switching through different VM sizes but it does not change the results.

**Comment by ryanberger-az** [comment]
Also I could be crazy but the VM size list looks... outdated. I remember this was more up to date with V3's and stuff and now they're gone. This seems to me to be all related to API calls to pull in these datasources.

![image](https://user-images.githubusercontent.com/60238330/142503126-74b2fccd-530a-4b49-a9e9-87ec3ce6eb3d.png)

**Comment by timja** [comment]
The VM size list has a hard coded fallback if it can’t load them

**Comment by ryanberger-az** [comment]
So what can I do to trouble shoot why the API calls are not functioning? Should this be an issue for another plugin? It pulls in the resource group, and I've reset the Managed Identity and started all over. It just wont query machine size or region, ever.

What plugin handles those calls?

**Comment by timja** [comment]
This plugin makes the call, using libraries from Azure SDK. No need to create another issue at this point

**Comment by ryanberger-az** [comment]
![image](https://user-images.githubusercontent.com/60238330/142642581-f3061f33-5491-4413-88a3-c91336e1d1ec.png)

It is pulling a token using a managed identity. I will try and use a service principal shortly.

**Comment by ryanberger-az** [comment]
@timja Are these error messages meaning that it's unable to acquire an access token? The text is a bit confusing. I am not sure if that is saying it is acquiring one, or it is retrying after expiration. Those are flooding my Jenkins system logs every 30 seconds.

![image](https://user-images.githubusercontent.com/60238330/142643530-afffbf41-fa0e-4e80-ac65-0dc157607b01.png)

**Comment by timja** [comment]
Don't think so but it shouldn't be doing that so often.
Are you on the latest version of the azure credentials plugin?

**Comment by ryanberger-az** [comment]
@timja Yea I am on the latest of all of the Azure plugins. I tried to roll back Azure credentials to test for you, but the VM Agent has a dependency on it so it won't let me do any testing if I downgrade Credentials.

![image](https://user-images.githubusercontent.com/60238330/142645089-d6da869e-31ec-41a7-a51e-d0b8f2f46920.png)

**Comment by ryanberger-az** [comment]
I am still digging into this and it is just getting more confusing. The error messages that Jenkins is spitting out in the Azure VM logs are literally just, not true. I can check the Managed Identity via IAM and see that the access its telling me it does not have, it does have. I have tried to cycle through toggling on/off Managed Identity to see if just _resetting it_ would fix it, but no luck. Same issues each time. The Managed Identity for the Jenkins Manager virtual machine properly has **Contributor** access, and this has worked without an issue for months for multiple environments that we have. I upgraded my dev environment to the newest version of the plugin to try and reproduce a colleague running into this same issue, and now, we both have environments that are not working with the same issues.

![image](https://user-images.githubusercontent.com/60238330/142676269-2e618a04-2e7e-49f4-9fa6-aa46e6850afe.png)

![image](https://user-images.githubusercontent.com/60238330/142676317-196329a0-d10b-4159-ae91-ace89b7ef909.png)

![image](https://user-images.githubusercontent.com/60238330/142676440-1cb97611-f96a-42cd-bc66-9b3035e72112.png)

![image](https://user-images.githubusercontent.com/60238330/142676496-8da09904-34ed-4b65-9dee-8d8f21d51a9e.png)

**Comment by ryanberger-az** [comment]
So,

Service principal works. I created a test one and changed my Cloud Config to use it. The drop downs immediately propagate and fill in with the API calls and proper responses. And, the deployments make it to Azure (and work). The Service Principal I created has the EXACT same level of access as the Managed Identity. Managed Identity was throwing incorrect errors about not having access, when it does have access upon examining it in IAM. I put two examples below. I strongly believe something with the newest update has negatively impacted the ability to use Managed Identity, and for us, Service Principal is not a long term option. We automate a deployment for customers and automate the creation of Managed Identity + assigning Contributor to the resource group.

@timja - can you please let me know if there is anything you can think of, or what I should try next? If you could please test with Managed Identity and let me know if you have success or not. This was not an issue before we updated to the latest VM Agent plugin. The only reason I updated was because a colleague reported this exact same issue after upgrading plugins to latest, so I updated to test if I had the same issues (and I did).

![image](https://user-images.githubusercontent.com/60238330/142683960-aed801d0-c40b-45fd-8fbd-f0408d932b66.png)

![image](https://user-images.githubusercontent.com/60238330/142683975-b1347a88-3fea-478e-a169-76220f744f4d.png)

![image](https://user-images.githubusercontent.com/60238330/142684028-8cc9a368-552b-41f7-9d61-eb80d8e74d1e.png)

![image](https://user-images.githubusercontent.com/60238330/142684052-bbdb0cec-6275-4f33-a9f9-778b36a48399.png)

**Comment by timja** [comment]
I'm setting up a test with MI currently

**Comment by timja** [comment]
I'm trying with a system assigned MI and verifying it just spins forever:

![image](https://user-images.githubusercontent.com/21194782/142684700-af069cf6-01df-40bb-8ae2-791b7b40bbc9.png)

**Comment by timja** [comment]
Looks to be https://github.com/Azure/azure-sdk-for-java/issues/25084

**Comment by timja** [comment]
Fixed in https://github.com/jenkinsci/azure-sdk-plugin/pull/73

Released in https://github.com/jenkinsci/azure-sdk-plugin/releases/tag/70.v63f6a95999a7
